### ISSUE TITLE
Suppression of E402 no longer works as expected.

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#365]
In GitLab by @hashbrowncipher on May 11, 2020, 18:27

Given a minimal Python script like:

```
import os

os.environ.get("PATH")  # noqa: E402

from io import BytesIO

os.write(1, (repr(BytesIO) + "\n").encode("ascii"))
```

Flake8 3.7.9 reported no issues (as desired). Flake8 3.8.0 reports `test.py:5:1: E402 module level import not at top of file`.

I'm not 100% sure, but I believe this is a flake8 (rather than pycodestyle) issue. The reason I think this is because the issue arises when I force a flake8 version change (without changing other deps). Diff of `pip-freeze`:

```
--- pip-old	2020-05-11 18:25:59.353774654 -0700
+++ pip-new	2020-05-11 18:26:15.999679319 -0700
@@ -1,5 +1,5 @@
 entrypoints==0.3
-flake8==3.7.9
+flake8==3.8.0
 mccabe==0.6.1
 pycodestyle==2.6.0
 pyflakes==2.2.0
```

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 18:32

this is an intentional side-effect of !331

prior to 3.8.0, plugins received a `noqa` attribute and could hide errors independent of the flake8 noqa parsing (leading to (for instance) `# noqa: ABC123` also erroneously hiding `E123` errors because `pycodestyle` received `noqa = True`)

this was changed in 3.8.0 (flake8 now always passes `noqa`) meaning that a `# noqa` comment can now only suppress errors on the line where the error appears

this change is also a precursor to "strict noqa" which I'm working on -- an option to make unused `noqa` an error (more details in #352)

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 18:32

the fix in your case is to put a `# noqa: E402` comment on the `from io import BytesIO` import

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 18:36

closed

**Comment by asottile** [comment]
In GitLab by @SeanMooney1 on May 12, 2020, 04:43

so !331 intentonally broke the api with out a major version bump by changing the fundemtal behaviour of where errors are reported.

if the fix is to change

```
os.environ.get("PATH")  # noqa: E402

from io import BytesIO
```

to 
```
os.environ.get("PATH")

from io import BytesIO # noqa: E402
```

that is effectivly redfining the meaning of E402

this is causing failure in our use of flake8 in our ci jobs in openstack due to similar cases where we now need to move # noqa commnets so i hope the desicsion to break user in a minor version was not taken lightly when !331 was merged.


by the way im not arguing that you shoudl not change change the behavior regarding passing noqa to plugins or not. im just saying that this is an api breakage and im not sure a minor version bump is really sufficient as its not backwards comatiable or additive so it really should have been a bump to 4.0.0

**Comment by asottile** [comment]
In GitLab by @ericvw on May 12, 2020, 06:21

It appears to be a behavior change that addresses bugs where errors could be hidden or being ignored on incorrect lines, as @asottile has indicated.  The error has been reporting for the correct line, but now the `# noqa: E402` is placed where "module level import not at top of fie" is being reported.

This is tough when fixing bugs that result in changing unintended, relied upon behavior.

**Comment by asottile** [comment]
In GitLab by @asottile on May 12, 2020, 07:27

@SeanMooney1 please consider your tone may be construed negatively, I'm doing this for free and a little empathy would be appreciated in these trying times (I was laid off due to the COVID shenanigans).  consider sponsoring the free software you're using.

___

that said, as is in line with our policies this was a minor change:

> New errors detected by dependencies

as such, no minor release of flake8 is "backward compatible" as we rev the minor revisions of the dependencies

**Comment by asottile** [comment]
In GitLab by @SeanMooney1 on May 12, 2020, 07:30

in nova we have some other examples

     e = self.assertRaises(
                    Exception, driver.VMwareVCDriver, None)  # noqa

need to be changed to 

     e = self.assertRaises(Exception, driver.VMwareVCDriver, None)  # noqa

this happens to fit under 80 columns but if it did not we woul have had to do

     e = self.assertRaises(Exception,
                           driver.VMwareVCDriver, None)  # noqa

and not do a hanging indent in that case.

we also have to change

 count = select([func.count()]).select_from(pci_devices).where(and_(
        pci_devices.c.deleted == 0,
        pci_devices.c.parent_addr == None,
        pci_devices.c.dev_type == 'type-VF')).execute().scalar()  # NOQA

to 

 count = select([func.count()]).select_from(pci_devices).where(and_(
        pci_devices.c.deleted == 0,
        pci_devices.c.parent_addr == None,  # noqa
        pci_devices.c.dev_type == 'type-VF')).execute().scalar()

we are not sure if sqlalcame is special casing 

  pci_devices.c.parent_addr == None,

so we cant jsut replace that with 

  pci_devices.c.parent_addr is None,

since i think its reflecting on the expression to generate the sql query.

so this does not just affect the case that was reported.

it move all warning for the end for multi line statements to the line in the multi line statemnet where the issue is.


in the grand scheme of things the fallout in nova so far is pretty minor for 3.8.0
but 3.8.1 has now adde a bunch more cases we have to deal with

nova/tests/unit/objects/test_objects.py:884:9: F811 redefinition of unused 'Parent' from line 876
./nova/tests/unit/api/openstack/compute/microversions.py:34:5: F811 redefinition of unused 'index' from line 29
./nova/tests/unit/api/openstack/compute/microversions.py:39:5: F811 redefinition of unused 'index' from line 34
./nova/tests/unit/api/openstack/compute/microversions.py:55:5: F811 redefinition of unused 'index' from line 49
./nova/tests/unit/api/openstack/compute/microversions.py:91:5: F811 redefinition of unused '_create' from line 86
./nova/api/openstack/compute/migrations.py:156:5: F811 redefinition of unused 'index' from line 148
./nova/api/openstack/compute/migrations.py:164:5: F811 redefinition of unused 'index' from line 156
./nova/api/openstack/compute/migrations.py:179:5: F811 redefinition of unused 'index' from line 164
./nova/api/openstack/compute/quota_sets.py:115:5: F811 redefinition of unused 'show' from line 109
./nova/api/openstack/compute/quota_sets.py:120:5: F811 redefinition of unused 'show' from line 115
./nova/api/openstack/compute/quota_sets.py:144:5: F811 redefinition of unused 'detail' from line 138
./nova/api/openstack/compute/quota_sets.py:149:5: F811 redefinition of unused 'detail' from line 144
./nova/api/openstack/compute/quota_sets.py:175:5: F811 redefinition of unused 'update' from line 168
./nova/api/openstack/compute/quota_sets.py:181:5: F811 redefinition of unused 'update' from line 175
./nova/api/openstack/compute/quota_sets.py:250:5: F811 redefinition of unused 'defaults' from line 244
./nova/api/openstack/compute/quota_sets.py:255:5: F811 redefinition of unused 'defaults' from line 250
./nova/api/openstack/compute/console_auth_tokens.py:79:5: F811 redefinition of unused 'show' from line 74
./nova/api/openstack/compute/hypervisors.py:213:5: F811 redefinition of unused 'index' from line 200
./nova/api/openstack/compute/hypervisors.py:219:5: F811 redefinition of unused 'index' from line 213
./nova/api/openstack/compute/hypervisors.py:245:5: F811 redefinition of unused 'detail' from line 232
./nova/api/openstack/compute/hypervisors.py:251:5: F811 redefinition of unused 'detail' from line 245
./nova/api/openstack/compute/hypervisors.py:303:5: F811 redefinition of unused 'show' from line 290
./nova/api/openstack/compute/services.py:383:5: F811 redefinition of unused 'update' from line 362
./nova/api/openstack/compute/instance_actions.py:80:5: F811 redefinition of unused '_get_instance' from line 76
./nova/api/openstack/compute/instance_actions.py:103:5: F811 redefinition of unused 'index' from line 86
./nova/api/openstack/compute/limits.py:57:5: F811 redefinition of unused 'index' from line 50
./nova/api/openstack/compute/limits.py:64:5: F811 redefinition of unused 'index' from line 57
./nova/api/openstack/compute/limits.py:71:5: F811 redefinition of unused 'index' from line 64
./nova/api/openstack/compute/keypairs.py:67:5: F811 redefinition of unused 'create' from line 48
./nova/api/openstack/compute/keypairs.py:88:5: F811 redefinition of unused 'create' from line 67
./nova/api/openstack/compute/keypairs.py:145:5: F811 redefinition of unused 'delete' from line 138
./nova/api/openstack/compute/keypairs.py:153:5: F811 redefinition of unused 'delete' from line 145
./nova/api/openstack/compute/keypairs.py:187:5: F811 redefinition of unused 'show' from line 179
./nova/api/openstack/compute/keypairs.py:193:5: F811 redefinition of unused 'show' from line 187
./nova/api/openstack/compute/keypairs.py:220:5: F811 redefinition of unused 'index' from line 213
./nova/api/openstack/compute/keypairs.py:228:5: F811 redefinition of unused 'index' from line 220
./nova/api/openstack/compute/keypairs.py:234:5: F811 redefinition of unused 'index' from line 228
./nova/api/openstack/compute/quota_classes.py:85:5: F811 redefinition of unused 'show' from line 80
./nova/api/openstack/compute/quota_classes.py:90:5: F811 redefinition of unused 'show' from line 85
./nova/api/openstack/compute/quota_classes.py:110:5: F811 redefinition of unused 'update' from line 104
./nova/api/openstack/compute/quota_classes.py:116:5: F811 redefinition of unused 'update' from line 110
./nova/api/openstack/compute/simple_tenant_usage.py:276:5: F811 redefinition of unused 'index' from line 269
./nova/api/openstack/compute/simple_tenant_usage.py:291:5: F811 redefinition of unused 'show' from line 284

again in all cases the # noqa
commnet used to go on the either aywaher on the statement or no the first decorator

but now i guess we need to move them all the the function line instead.

https://review.opendev.org/#/c/727140/1 is attemting to adress the issue in nova but this will effect every openstack project to some degree

**Comment by asottile** [comment]
In GitLab by @asottile on May 12, 2020, 07:36

nothign really changed between 3.8.0 and 3.8.1 so I suspect you've got other issues there - F811 also didn't change between flake8 3.7 and 3.8 so I'm not sure what's up with that

it's reasonably expected that across minor versions of flake8 / pyflakes / pycodestyle you will need to make small changes to extraordinary code

**Comment by asottile** [comment]
In GitLab by @SeanMooney1 on May 12, 2020, 08:10

@asottile sorry it wasnt ment to be a personal dig.
i read the tone of your intial comment "this is an intentional side-effect of !331"
as a slightly flipent disregard for the consumer of flake8 

i was also irritated by the fact that the inital patch to "fix" this in nova did not explain why the individual changes where being made and i had been trying to get stephen to add comment  for each of # noqa ro at least do # noqa: <rule> so that we could follow why it was done.

i didn't intend it to filter true into the comment i left.

and yes we normally do make minor chnages 

going back as far as 3.5 i am seeing those error locally on pyton 3.8 but i know we used to have 3.6 as our min version last release or the one before so im wondering if this behavior change is unrelateed to the flake8 release and perhaps due a change in one of the depencies. ill see if i see the same behavior im seeing on ubuntu 20.04 on 18.04 because this should not have been passing based on what im seeing locally.

**Comment by asottile** [comment]
In GitLab by @csala on May 19, 2020, 02:28

Following up from here: https://github.com/PyCQA/pycodestyle/issues/939#issuecomment-630382520

@asottile I have to agree with @SeanMooney1 that this is effectively breaking the API and should have been pushed to a major version bump instead of a minor.

Putting the version change aside, is there any way to disable this behavior and go back to the old one?

Simply putting the `noqa` explicitly on every other affected line is not the way to go. What if there is a module that has 1 exceptional line of code at the top line and then 30 lines of regular imports? Is there really no way to ignore the E402 errors other than setting a `noqa` on every line?
And per-file ignores do no help either: what happens if I want to validate E402 on the res of the file and only ignore the 30 imports at the top?

So, in other words, is there any way to completely "hide" a line from the validation instead of just having to alter everything else because of that line?

**Comment by asottile** [comment]
In GitLab by @asottile on May 19, 2020, 04:20

locked this issue
